File: interfaces/cli/gateway_client.py
"""CLI Gateway /v1 client (DDE-056 client-parity surface).

Mirrors ``interfaces/dashboard/static/gateway.js`` and Android
``GatewayAllowlist`` — the same six endpoints, no invented lists or
streams. The Stage-1 ``dde mission create|status|trace`` commands still
call ``engine.*`` directly; this module is the Gateway path the golden
CLI/web/Android parity fixture exercises so all three clients share one
authoritative control plane (Ch.15.1/15.2).
"""

from __future__ import annotations

from datetime import UTC, datetime
from typing import Any, Final
from uuid import uuid4

import httpx

#: Must stay identical to dashboard ``GatewayApiClient.ALLOWED_PATHS`` and
#: Android ``GatewayAllowlist.ALLOWED_PATHS``.
ALLOWED_PATHS: Final[tuple[str, ...]] = (
    "POST /v1/sessions",
    "POST /v1/sessions/{id}/resume",
    "POST /v1/sessions/{id}/close",
    "POST /v1/commands",
    "GET /v1/missions/{id}",
    "GET /v1/mission-control/{id}",
)

#: Human operator scopes used by dashboard ``app.js`` and Android
#: ``HumanMissionScopes`` for Mission Control.
HUMAN_MISSION_SCOPES: Final[tuple[str, ...]] = (
    "mission.read",
    "mission.create",
    "mission.control",
    "approval.read",
    "approval.decide",
    "approval.request",
    "credential.capture",
)


class GatewayClient:
    """Thin async client over the allowlisted Gateway surface."""

    def __init__(self, http: httpx.AsyncClient, *, base_path: str = "/v1") -> None:
        self._http = http
        self._base = base_path.rstrip("/")

    async def open_session(
        self,
        *,
        principal_id: str,
        client_type: str = "human",
        scopes: list[str] | tuple[str, ...] | None = None,
        subscriptions: list[str] | None = None,
    ) -> dict[str, Any]:
        response = await self._http.post(
            f"{self._base}/sessions",
            json={
                "principal_id": principal_id,
                "client_type": client_type,
                "protocol_version": "1",
                "scopes": list(scopes if scopes is not None else HUMAN_MISSION_SCOPES),
                "subscriptions": subscriptions
                if subscriptions is not None
                else ["mission"],
            },
        )
        response.raise_for_status()
        return response.json()

    async def resume_session(
        self, session_id: str, *, last_event_at: str | None = None
    ) -> dict[str, Any]:
        response = await self._http.post(
            f"{self._base}/sessions/{session_id}/resume",
            json={"last_event_at": last_event_at},
        )
        response.raise_for_status()
        return response.json()

    async def close_session(self, session_id: str) -> dict[str, Any]:
        response = await self._http.post(
            f"{self._base}/sessions/{session_id}/close",
            json={},
        )
        response.raise_for_status()
        return response.json()

    async def accept_command(
        self,
        *,
        session_id: str,
        principal_id: str,
        command_type: str,
        target_type: str,
        target_id: str,
        parameters: dict[str, Any],
        idempotency_key: str,
        command_id: str | None = None,
    ) -> tuple[int, dict[str, Any]]:
        """POST /v1/commands. Returns ``(http_status, body)``.

        Callers must treat 202 as acceptance only (Ch.15.2) and re-read
        authoritative state by id.
        """
        body = {
            "command_id": command_id or str(uuid4()),
            "idempotency_key": idempotency_key,
            "principal_id": principal_id,
            "client_session_id": session_id,
            "target_type": target_type,
            "target_id": target_id,
            "command_type": command_type,
            "parameters": parameters,
            "requested_at": datetime.now(UTC).isoformat(),
            "protocol_version": "1",
        }
        response = await self._http.post(f"{self._base}/commands", json=body)
        return response.status_code, response.json()

    async def read_mission(
        self, *, session_id: str, principal_id: str, mission_id: str
    ) -> dict[str, Any]:
        response = await self._http.get(
            f"{self._base}/missions/{mission_id}",
            headers={
                "X-Session-Id": session_id,
                "X-Principal-Id": principal_id,
            },
        )
        response.raise_for_status()
        return response.json()

    async def read_mission_control(
        self, *, session_id: str, principal_id: str, mission_id: str
    ) -> dict[str, Any]:
        response = await self._http.get(
            f"{self._base}/mission-control/{mission_id}",
            headers={
                "X-Session-Id": session_id,
                "X-Principal-Id": principal_id,
            },
        )
        response.raise_for_status()
        return response.json()

File: interfaces/dde-studio/shared/gatewayClient.ts
/**
 * Typed client for the live Gateway /v1 surface (engine/gateway/api.py):
 * session open/resume/close, command acceptance (202), mission read and
 * mission-control projection. Shapes mirror engine/contracts (generated
 * from schemas/) — do not widen them client-side.
 *
 * Not yet served by Core: mission/run/event LIST endpoints (DDE-027).
 * Fleet-room list views therefore stay empty; this client only calls
 * endpoints that really exist. Errors map onto the Chapter 15.5 Error
 * contract (error family + retryable flag).
 */

export interface GatewayMission {
  mission_id: string;
  tenant_id: string;
  project_id: string;
  slug: string;
  title: string;
  intent: string;
  success_definition: string;
  scope: string[];
  requirement_refs: string[];
  status:
    | "CREATED"
    | "ACTIVE"
    | "PARTIAL"
    | "PAUSED"
    | "COMPLETED"
    | "FAILED"
    | "CANCELLED";
  autonomy_ceiling: number;
  lock_version: number;
  created_at: string;
  updated_at: string;
}

export interface GatewayMissionControl {
  mission_id: string;
  tenant_id: string;
  project_id: string;
  slug: string;
  title: string;
  status: GatewayMission["status"];
  autonomy_ceiling: number;
  lock_version: number;
  task_total: number;
  task_counts: Record<string, number>;
  tasks_completed: number;
  open_attention_items: number;
  attention_debt: number;
  human_minutes: number;
  approvals_per_mission: number;
  approvals_by_type: Record<string, number>;
  blocked_requests: number;
  standing_approval_usage: number;
  last_event_at?: string | null;
  created_at: string;
  updated_at: string;
}

export interface GatewaySession {
  session_id: string;
  tenant_id: string;
  principal_id: string;
  client_type: string;
  device_id?: string | null;
  protocol_version: string;
  scopes: string[];
  connected_at: string;
  last_seen_at: string;
  subscriptions: string[];
  status: string;
  created_at: string;
  updated_at: string;
}

export interface CommandAcceptance {
  command_id: string;
  status: string;
  target_type: string;
  target_id: string;
  payload: Record<string, unknown>;
}

/** Chapter 15.5 Error contract subset clients act on. */
export class GatewayApiError extends Error {
  constructor(
    readonly errorFamily: string,
    readonly detail: string,
    readonly httpStatus: number,
    readonly retryable?: boolean,
  ) {
    super(`Gateway ${errorFamily}: ${detail} (HTTP ${httpStatus})`);
    this.name = "GatewayApiError";
  }
}

export interface OpenSessionInput {
  principalId: string;
  clientType: string;
  scopes: string[];
  deviceId?: string;
  subscriptions?: string[];
  protocolVersion?: string;
}

export class GatewayApiClient {
  constructor(private readonly baseUrl: string) {}

  getBasePath(): string {
    return `${this.baseUrl}/v1`;
  }

  async openSession(input: OpenSessionInput): Promise<GatewaySession> {
    return this.post("/sessions", {
      principal_id: input.principalId,
      client_type: input.clientType,
      device_id: input.deviceId ?? null,
      protocol_version: input.protocolVersion ?? "1",
      scopes: input.scopes,
      subscriptions: input.subscriptions ?? [],
    });
  }

  async resumeSession(
    sessionId: string,
    lastEventAt?: string,
  ): Promise<{
    session: GatewaySession;
    fresh_snapshot: boolean;
    events: unknown[];
  }> {
    return this.post(`/sessions/${sessionId}/resume`, {
      last_event_at: lastEventAt ?? null,
    });
  }

  async closeSession(sessionId: string): Promise<GatewaySession> {
    return this.post(`/sessions/${sessionId}/close`, {});
  }

  async acceptCommand(command: {
    commandId: string;
    idempotencyKey: string;
    principalId: string;
    clientSessionId: string;
    targetType: string;
    targetId: string;
    commandType: string;
    parameters: Record<string, unknown>;
    protocolVersion?: string;
  }): Promise<CommandAcceptance> {
    return this.post("/commands", {
      command_id: command.commandId,
      idempotency_key: command.idempotencyKey,
      principal_id: command.principalId,
      client_session_id: command.clientSessionId,
      target_type: command.targetType,
      target_id: command.targetId,
      command_type: command.commandType,
      parameters: command.parameters,
      requested_at: new Date().toISOString(),
      protocol_version: command.protocolVersion ?? "1",
    });
  }

  async readMission(
    sessionId: string,
    principalId: string,
    missionId: string,
  ): Promise<GatewayMission> {
    return this.get(`/missions/${missionId}`, sessionId, principalId);
  }

  async readMissionControl(
    sessionId: string,
    principalId: string,
    missionId: string,
  ): Promise<GatewayMissionControl> {
    return this.get(`/mission-control/${missionId}`, sessionId, principalId);
  }

  async readFrontendSnapshot(
    sessionId: string,
    principalId: string,
    missionId: string,
  ): Promise<Record<string, unknown>> {
    return this.get(
      `/missions/${missionId}/frontend/snapshot`,
      sessionId,
      principalId,
    );
  }

  async readDdeChat(
    sessionId: string,
    principalId: string,
    missionId: string,
  ): Promise<Record<string, unknown>> {
    return this.get(`/missions/${missionId}/chat/latest`, sessionId, principalId);
  }

  async readDdeChatResource(
    sessionId: string,
    principalId: string,
    missionId: string,
    suffix: string,
    query?: URLSearchParams,
  ): Promise<Record<string, unknown>> {
    const clean = suffix.replace(/^\/+/, "");
    const tail = query && [...query.keys()].length ? `?${query.toString()}` : "";
    let path: string;
    if (clean.startsWith("?")) {
      path = `/missions/${missionId}/chat/conversations${clean}`;
    } else if (clean === "models") {
      path = `/missions/${missionId}/chat/models${tail}`;
    } else if (clean.length) {
      path = `/missions/${missionId}/chat/conversations/${clean}${tail}`;
    } else {
      path = `/missions/${missionId}/chat/conversations${tail}`;
    }
    return this.get(path, sessionId, principalId);
  }

  async uploadDdeChatAttachment(
    sessionId: string,
    principalId: string,
    missionId: string,
    conversationId: string,
    attachmentId: string,
    bytes: Uint8Array,
    idempotencyKey: string,
  ): Promise<Record<string, unknown>> {
    const response = await fetch(
      `${this.getBasePath()}/missions/${missionId}/chat/conversations/${conversationId}/attachments/${attachmentId}/content`,
      {
        method: "PUT",
        headers: {
          Accept: "application/json",
          "Content-Type": "application/octet-stream",
          "X-Session-Id": sessionId,
          "X-Principal-Id": principalId,
          "X-Idempotency-Key": idempotencyKey,
        },
        body: bytes,
      },
    );
    return this.parse<Record<string, unknown>>(response);
  }

  /** @deprecated Compatibility alias for pre-universal Frontend Studio Chat. */
  async readFrontendChat(
    sessionId: string,
    principalId: string,
    missionId: string,
  ): Promise<Record<string, unknown>> {
    return this.readDdeChat(sessionId, principalId, missionId);
  }

  /** @deprecated Compatibility alias for pre-universal Frontend Studio Chat. */
  async readFrontendChatResource(
    sessionId: string,
    principalId: string,
    missionId: string,
    suffix: string,
    query?: URLSearchParams,
  ): Promise<Record<string, unknown>> {
    return this.readDdeChatResource(sessionId, principalId, missionId, suffix, query);
  }

  /** @deprecated Compatibility alias for pre-universal Frontend Studio Chat. */
  async uploadFrontendChatAttachment(
    sessionId: string,
    principalId: string,
    missionId: string,
    conversationId: string,
    attachmentId: string,
    bytes: Uint8Array,
    idempotencyKey: string,
  ): Promise<Record<string, unknown>> {
    return this.uploadDdeChatAttachment(
      sessionId, principalId, missionId, conversationId, attachmentId, bytes, idempotencyKey,
    );
  }

  async readFrontendAudit(
    sessionId: string,
    principalId: string,
    missionId: string,
    suffix = "summary",
  ): Promise<Record<string, unknown>> {
    const clean = suffix.replace(/^\/+/, "");
    return this.get(
      `/missions/${missionId}/frontend/audit/${clean}`,
      sessionId,
      principalId,
    );
  }

  async readFrontendSources(
    sessionId: string,
    principalId: string,
    missionId: string,
  ): Promise<Record<string, unknown>> {
    return this.get(`/missions/${missionId}/frontend/sources`, sessionId, principalId);
  }

  async readFrontendSourceArtifact(
    sessionId: string,
    principalId: string,
    missionId: string,
    artifactId: string,
  ): Promise<Record<string, unknown>> {
    return this.get(
      `/missions/${missionId}/frontend/sources/artifacts/${artifactId}`,
      sessionId,
      principalId,
    );
  }

  async readFrontendSourceProvenance(
    sessionId: string,
    principalId: string,
    missionId: string,
    subjectKind: string,
    subjectRef: string,
  ): Promise<Record<string, unknown>> {
    const query = new URLSearchParams({ subject_kind: subjectKind, subject_ref: subjectRef });
    return this.get(
      `/missions/${missionId}/frontend/sources/provenance?${query.toString()}`,
      sessionId,
      principalId,
    );
  }

  async readFrontendSourceTargetBlend(
    sessionId: string,
    principalId: string,
    missionId: string,
    scopeKey: string,
  ): Promise<Record<string, unknown>> {
    const query = new URLSearchParams({ scope_key: scopeKey });
    return this.get(
      `/missions/${missionId}/frontend/sources/target-blend?${query.toString()}`,
      sessionId,
      principalId,
    );
  }

  async readFrontendCandidateScore(
    sessionId: string,
    principalId: string,
    missionId: string,
    candidateId: string,
  ): Promise<Record<string, unknown>> {
    return this.get(
      `/missions/${missionId}/frontend/sources/candidates/${candidateId}/score`,
      sessionId,
      principalId,
    );
  }

  async readFrontendPreview(
    sessionId: string,
    principalId: string,
    missionId: string,
    previewSessionId: string,
  ): Promise<Record<string, unknown>> {
    return this.get(
      `/missions/${missionId}/frontend/previews/${previewSessionId}`,
      sessionId,
      principalId,
    );
  }

  async readFrontendInspector(
    sessionId: string,
    principalId: string,
    missionId: string,
    candidateId: string,
    pxgKey: string,
  ): Promise<Record<string, unknown>> {
    const query = new URLSearchParams({ pxg_key: pxgKey });
    return this.get(
      `/missions/${missionId}/frontend/inspector/${candidateId}?${query.toString()}`,
      sessionId,
      principalId,
    );
  }

  private async get<T>(
    path: string,
    sessionId: string,
    principalId: string,
  ): Promise<T> {
    const response = await fetch(`${this.getBasePath()}${path}`, {
      method: "GET",
      headers: {
        Accept: "application/json",
        "X-Session-Id": sessionId,
        "X-Principal-Id": principalId,
      },
    });
    return this.parse<T>(response);
  }

  private async post<T>(path: string, body: unknown): Promise<T> {
    const response = await fetch(`${this.getBasePath()}${path}`, {
      method: "POST",
      headers: {
        "Content-Type": "application/json",
        Accept: "application/json",
      },
      body: JSON.stringify(body),
    });
    return this.parse<T>(response);
  }

  private async parse<T>(response: Response): Promise<T> {
    if (!response.ok) {
      let family = `HTTP_${response.status}`;
      let detail = response.statusText;
      let retryable: boolean | undefined;
      try {
        const body = (await response.json()) as {
          error_code?: string;
          message?: string;
          retryable?: boolean;
        };
        if (body.error_code) {
          family = body.error_code;
        }
        if (body.message) {
          detail = body.message;
        }
        if (typeof body.retryable === "boolean") {
          retryable = body.retryable;
        }
      } catch {
        // Non-JSON error body: fall back to status-line detail.
      }
      throw new GatewayApiError(family, detail, response.status, retryable);
    }
    return (await response.json()) as T;
  }
}
